### PR TITLE
Implemented Future Effective Validators

### PR DESCRIPTION
**Pull request by cameron-freshworks**
*Issue #:* /bcgov/entity#2131

*Description of changes:*
* Implemented validators an effective date for future effective incorporation filings.
* Updated unit testing.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 1]
```diff
@@ -186,3 +192,40 @@ def validate_share_structure(incorporation_json) -> Error:
...
+
+    Rules:
+        - The effective date must be the correct format.
+        - The effective date must be a minimum of 2 minutes in the future.
+        - The effective date must be a maximum of 10 days in the future.
+    """
+    # Setup
+    msg = []
+    now = dt.utcnow()
+    now_plus_2_minutes = now + timedelta(minutes=2)
```
Comment (lekshmimallika-aot) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: Is this working fine if you choose exactly 2 mins in the UI?

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 2]
```diff
@@ -186,3 +192,40 @@ def validate_share_structure(incorporation_json) -> Error:
...
+    now_plus_10_days = now + timedelta(days=10)
+
+    try:
+        filing_effective_date = incorporation_json['filing']['header']['effectiveDate']
```
Comment (lekshmimallika-aot) [inline legal-api/src/legal_api/services/filings/validations/incorporation_application.py:212]: Are we passing effective date in case of immediate filing? Is the validation passing for immediate filing

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: The front end won't let you File And Pay if it is any sooner than 2 minutes, yea.

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 2 above)
Comment (lekshmimallika-aot) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:212]: Also I think the legal API returns the effective date in the filing header when filing is resumed ( after save). It would be a good check to see whether these validations are passing in the second submit

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: For example, if you select exactly 2 minutes ahead, but take 5 minutes to file. You will have to update that time before filing and paying.

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 2 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:212]: So, correct me if i'm wrong, but if it is immediate, effective date is being passed as `null` or `None` to the backend. This check should return the validations with no errors if that is the case. It is the second test scenario. ( At least those were my intentions :) )

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:212]: oh interesting! How could i go about testing that..

File: legal-api/tests/unit/api/test_business_incorporation.py [hunk 3]
```diff
@@ -66,51 +75,54 @@ def test_post_new_incorporation(session, client, jwt):
...
-    assert filing_id == rv.json['filing']['header']['filingId']
+        assert HTTPStatus.OK == rv.status_code
+        assert filing_id == rv.json['filing']['header']['filingId']
 
 
 def test_put_draft_incorporation_filing(session, client, jwt):
     """Assert that an incorporation filing can be put (updated) to filings endpoint."""
-    nr_number = 'NR 1234567'
     filing_json = copy.deepcopy(INCORPORATION_FILING_TEMPLATE)
     filing_json['filing']['incorporationApplication']['nameRequest']['nrNumber'] = nr_number
```
Comment (kialj876) [inline legal-api/tests/unit/api/test_business_incorporation.py:127]: How is nr_number getting assigned now? It looks like this should cause an error to me, am I missing something?

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (lekshmimallika-aot) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: utcnow will always be ahead of the submission time by milliseconds or so due to network latency and stuff right? I think there is achance that exact 2 mins from the UI may cause a validation failure. Could you please test it out from UI

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: Good point for sure, i know this is an extreme edge case ( someone paying $100 to have it effective in 2 mins from now 😆 ), so should we adjust the time and give a buffer on the back end. It could potentially pass front end validation and then fail the validator during the request.

The front end will not make the request if it is < 2 mins, but you're right, we should take into account the latency, would adding an extra minute be acceptable?

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py [hunk 4]
```diff
@@ -186,3 +192,40 @@ def validate_share_structure(incorporation_json) -> Error:
...
+    except KeyError:
+        return msg
+
+    try:
+        effective_date = dt.fromisoformat(filing_effective_date)
+    except ValueError:
+        msg.append({'error': babel('%s is an invalid ISO format for effective_date.') % filing_effective_date})
+        return msg
+
+    if effective_date < now_plus_2_minutes:
```
[the remark this answers is not in the document]
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:222]: The front end actually stores and makes the request with `effectiveDate` it in UTC('IsoString').
 Just from working with DateTimes earlier, we knew that since the backend was handling it in this format, it was simpler to avoid the time conversions. We just convert convert the DateTime when presenting it in the UI.

File: legal-api/tests/unit/api/test_business_incorporation.py (hunk 3 above)
Comment (cameron-freshworks) [reply legal-api/tests/unit/api/test_business_incorporation.py:127]: I defined `nr_number` as the `indentifier` at the beginning of the test setup, and assigned that identifier to the filing nrNumber in each test. Basically just defined the NR once opposed to defining it in each test. 

see line 31 for the definition

an example of how i assigned it in each test.
`f['filing']['incorporationApplication']['nameRequest']['nrNumber'] = identifier`

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kialj876) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:222]: awesome

File: legal-api/tests/unit/api/test_business_incorporation.py (hunk 3 above)
Comment (kialj876) [reply legal-api/tests/unit/api/test_business_incorporation.py:127]: Oh cool! Nice

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (lekshmimallika-aot) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: Also, this particular one need not block the PR as such and the fix (if any) can be a separate PR

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 2 above)
Comment (lekshmimallika-aot) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:212]: Try 'save and resume' for an immediate filing. In the second submission(dont change anything just click on file and pay ) if it passes, it is fine

Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:212]: kk, will do thanks!

File: legal-api/src/legal_api/services/filings/validations/incorporation_application.py (hunk 1 above)
Comment (cameron-freshworks) [reply legal-api/src/legal_api/services/filings/validations/incorporation_application.py:208]: Absolutely, will look into this, thanks Lekshmi!
